File: observer/models/article.py
from __future__ import annotations

import datetime
from dataclasses import dataclass

from .summary import Summary


@dataclass
class Article:
    _id: str
    title: str
    pub_date: datetime.datetime
    author: str
    summary: Summary

    @property
    def url(self) -> str:
        return self._id

    @classmethod
    def from_dict(cls, d: dict) -> Article:
        return Article(
            _id=d["id"],
            title=d["title"],
            # Python 3.9's fromisoformat cannot parse the RFC 3339 "Z" suffix.
            pub_date=datetime.datetime.fromisoformat(d["pub_date"].replace("Z", "+00:00")),
            author=d["author"],
            summary=Summary.from_dict(d["summary"]),
        )

File: go/internal/entities/article.go
package entities

import "time"

type Article struct {
	ID      string    `json:"id" bson:"_id"`
	Title   string    `json:"title" bson:"title"`
	PubDate time.Time `json:"pub_date" bson:"pub_date"`
	Author  string    `json:"author" bson:"author"`
	Summary *Summary  `json:"summary" bson:"summary"`
}

func (a *Article) URL() string {
	return a.ID
}
